### ISSUE TITLE
Compile to WASM

### ISSUE THREAD

**Issue by imjasonh** [ko-build/ko#675]
Go can compile to WASM (`GOOS=js GOARCH=wasm`), and WASM can be packaged into images and executed on K8s clusters using [krustlet](https://krustlet.dev): https://developer.okta.com/blog/2022/01/28/webassembly-on-kubernetes-with-rust

So, maybe `ko` can build a WASM executable, and put it in a multiarch image, alongside the other ko-built archs.

This wouldn't work for Go executables that need to call into their base images (I don't think WASM images have multiple layers, and the environment isn't very linux-ey...yet?), but it should mostly maybe kinda work if they're just static binaries, which is exactly ko's sweet spot.

An example WASM image that kruslet can apparently run: https://explore.ggcr.dev/?image=webassembly.azurecr.io%2Fhello-wasm%3Av1 -- note media types and the annotation.

Questions:
- What OCI platform does kruslet expect when it's given a multiplatform image? `js/wasm`?

- How should users express this? `ko build ./ --wasm` to additionally produce a WASM image in the index?

**Comment by imjasonh** [comment]
> * What OCI platform does kruslet expect when it's given a multiplatform image? `js/wasm`?

Kruslet doesn't appear to support pulling from indexes at all, only OCI artifacts with the WASM media types. So this is probably not going to be something worth shoehorning into ko.

This might still make sense as a separate Go-specific pipeline for building WASM images, a la [wasm-to-oci](https://github.com/engineerd/wasm-to-oci) with fewer steps, but it won't coexist alongside regular OCI images.

**Comment by imjasonh** [comment]
Reopening as I still think this is an idea worth pulling on.

**Comment by imjasonh** [comment]
cf: Istio's WASM plugins are also packaged as OCI images ([building tutorial](https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/how-to-build-oci-images.md), [spec](https://github.com/solo-io/wasm/blob/master/spec/spec-compat.md))

TL;DR: In this case, the format seems to be:

- the image must be a single-arch OCI image (not index)
- the image must contain one layer
- the layer must contain a file named `plugin.wasm` (entrypoint is ignored)
- and optionally a `runtime-config.json`

This seems fine for Istio, there's not a lot of use cases for multi-platform WASM and non-WASM plugins there I think, but it's something to consider as folks have gone down this path before and made certain decisions.

**Comment by imjasonh** [comment]
ref: Similar discussion in buildpacks, to allow buildpacks to run and produce a wasm image instead of the standard OCI image: https://github.com/buildpacks/lifecycle/issues/820

In this case, it would be either-or, and buildpacks wouldn't create a single multi-platform index containing the wasm image and the standard linux/amd64 (e.g.) image. But maybe if that was supported somewhere, they could.
